application-templates/starter-typescript/src/helpers.ts

```diff
@@ -0,0 +1,75 @@
+import {
+  transformLocalizedStringToLocalizedField,
+  transformLocalizedFieldToLocalizedString,
+} from '@commercetools-frontend/l10n';
+import { ApolloError, type ServerError } from '@apollo/client';
+import type { TChannel } from '../types/generated/ctp';
+
+export const getErrorMessage = (error: ApolloError) =>
+  error.graphQLErrors?.map((e) => e.message).join('\n') || error.message;
+
+const isServerError = (
+  error: ApolloError['networkError']
+): error is ServerError => {
+  return Boolean((error as ServerError)?.result);
+};
+
+export const extractErrorFromGraphQlResponse = (graphQlResponse: unknown) => {
+  if (graphQlResponse instanceof ApolloError) {
+    if (
+      isServerError(graphQlResponse.networkError) &&
+      graphQlResponse.networkError?.result?.errors.length > 0
+    ) {
+      return graphQlResponse?.networkError?.result.errors;
+    }
+
+    if (graphQlResponse.graphQLErrors?.length > 0) {
+      return graphQlResponse.graphQLErrors;
+    }
+  }
+
+  return graphQlResponse;
+};
+
+type LocalizedField = {
+  locale: string;
+  value: string;
+};
+
+type TChangeNameActionPayload = Record<'name', LocalizedField>;
+type SyncAction = { action: string; [x: string]: unknown };
+type GraphqlUpdateAction = Record<string, Record<string, unknown>>;
+
+const getNameFromPayload = (payload: TChangeNameActionPayload) => ({
+  name: transformLocalizedStringToLocalizedField(payload.name),
+});
+
+const hasActionPayloadNameProperty = (
+  actionPayload: Record<string, unknown>
+): actionPayload is TChangeNameActionPayload => {
+  return Boolean((actionPayload as TChangeNameActionPayload)?.name);
+};
+
+const convertAction = (
+  actionName: SyncAction['action'],
+  actionPayload: Record<string, unknown>
+) => ({
+  [actionName]:
+    actionName === 'changeName' && hasActionPayloadNameProperty(actionPayload)
+      ? getNameFromPayload(actionPayload)
+      : actionPayload,
+});
```

Comment:
Thanks, nice idea with changing the signature 👍
Updated in 24b692b9986c2a706de226d9a2ee0114c48c7522

Again I would argue that using `isChangeNameActionPayload` (I changed the name) type-guard, is a "slightly type-safer" approach than a type assertion, so I left that unchanged but I'm fine with changing that if needed. Personally I feel that type-guards may be used only once.

Reply:
Yes, I get your point.

So the goal here would be to be sure that we only call the `getNameFromPayload` function if the received action is a `changeName` typed one.
We already know that because of the received `actionName` so we also want to make sure that an action of type `actionName` has the expected payload just in case someone created an invalid action (the right name but the wrong payload).
So, the expected payload is defined for what we are going to do with it in the `getNameFromPayload` function, as that's the one which will receive the payload to be transformed.
That function uses `transformLocalizedStringToLocalizedField` internally which and that function expects [only one argument of type `LocalizedString`](https://github.com/commercetools/merchant-center-application-kit/blob/main/packages/l10n/src/localize.ts#L36) but, the payload we have defined for the `TChangeNameActionPayload` is of type `LocalizedField` instead.

So, either we are using the wrong function (and we should use `transformLocalizedFieldToLocalizedString` instead), or we have not chosen the right type for the payload in the `TChangeNameActionPayload` (and it should be `LocalizedString`).

At the end, what we are talking about here is how to guard the call to the transformation of the localized field in the payload for `changeName` actions convertion.

Reply:
That is actually a nice catch 👍 It should be `LocalizedString` and it type-checked before because `LocalizedField` is subtype of `LocalizedString` (which is basically an object type).

> At the end, what we are talking about here is how to guard the call to the transformation of the localized field in the payload for changeName actions convertion.

Agreed. 

I would still personally go for the type guard, but `... Boolean(actionPayload.name)
        ? getNameFromPayload(actionPayload as TChangeNameActionPayload) ...` makes just as much sense, and I'm fine with that too